src/pie_utils/sequence_tagging/ill_formed.py

```diff
@@ -0,0 +1,472 @@
+                    # if tag sequence starts with L tag, convert L to U, e.g: O O La *a .. to O O Ua *a ..
+                    new_tag_sequence.append(f"U-{current_tag_type}")
+                else:
+                    label = new_tag_sequence[i]
+            # Note : It cannot be the case that we encounter a B while traversing backward as this situation would
+            # already be handled in process_B
+            if label[0] in ["L", "U"]:
+                # if L or U is encountered
+                tag_type = label.partition("-")[2]
+                if tag_type == current_tag_type:
+                    # La O La or Ua O La
+                    if label[0] == "L":
+                        # LOL to OOL
+                        new_tag_sequence[-count] = "O"
+                        new_tag_sequence.append(f"L-{current_tag_type}")
+                    else:
+                        # UOL to BOL
+                        new_tag_sequence[-count] = f"B-{current_tag_type}"
+                        new_tag_sequence.append(f"L-{current_tag_type}")
+                else:
+                    if count == 1:
+                        # Ua Lb to Ua Ub
+                        new_tag_sequence.append(f"U-{current_tag_type}")
+                    else:
+                        # La O Lb to La Bb Lb and Ua O Lb to Ua Bb Lb
+                        count -= 1
+                        new_tag_sequence[-count] = f"B-{current_tag_type}"
+                        new_tag_sequence.append(f"L-{current_tag_type}")
+        else:
+            if label != "O":
+                raise InvalidTagSequence(tag_sequence)
+            else:
+                new_tag_sequence.append(label)
+        index += 1
+    return new_tag_sequence
+
+
+def fix_bio(tag_sequence: List[str]):
+    """Given a sequence corresponding to BIO tags, extracts spans. Spans are inclusive and can be
+    of zero length, representing a single word span. Ill-formed spans are also included (i.e those
+    which do not start with a "B-LABEL"), as otherwise it is possible to get a perfect precision
+    score whilst still predicting ill-formed spans in addition to the correct spans. This function
+    works properly when the spans are unlabeled (i.e., your labels are simply "B", "I", and "O").
+
+    # Parameters
+    tag_sequence : `List[str]`, required.
+        The integer class labels for a sequence.
+    classes_to_ignore : `List[str]`, optional (default = `None`).
+        A list of string class labels `excluding` the bio tag
+        which should be ignored when extracting spans.
+    include_ill_formed: `bool`, optional (default = `True`).
+        If this flag is enabled, include spans that do not start with "B". Otherwise, these are ignored.
+    # Returns
+    spans : `List[TypedStringSpan]`
+        The typed, extracted spans from the sequence, in the format (label, (span_start, span_end)).
+        Note that the label `does not` contain any BIO tag prefixes.
+    """
```

Comment:
I don't think this is valid anymore, please update.

Reply:
Yes, documentation wasn't updated, will start with it now.

Reply:
ok, no problem :)

Reply:
Documentation added [16e7b38](https://github.com/ArneBinder/pie-utils/pull/28/commits/16e7b38a2f354960512465bee3a4e526283feed3) and [83fadd7](https://github.com/ArneBinder/pie-utils/pull/28/commits/83fadd7ef2d512ef7e7dadece536d2024f5af4ca)

Reply:
This looks still not right.

Reply:
Now it should be fixed in [0683266](https://github.com/ArneBinder/pie-utils/pull/28/commits/0683266650489383033c50341f2108d93ce1276e)